server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -9,35 +9,55 @@
 package org.elasticsearch.search.fetch.subphase;
 
 import org.apache.lucene.index.LeafReaderContext;
-import org.elasticsearch.common.document.DocumentField;
-import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.index.mapper.IgnoredFieldMapper;
+import org.elasticsearch.index.mapper.LegacyTypeFieldMapper;
+import org.elasticsearch.index.mapper.RoutingFieldMapper;
 import org.elasticsearch.search.fetch.FetchContext;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
 import org.elasticsearch.search.fetch.StoredFieldsSpec;
 
 import java.io.IOException;
-import java.util.Map;
+import java.util.Collections;
+import java.util.List;
 
 /**
  * A fetch sub-phase for high-level field retrieval. Given a list of fields, it
  * retrieves the field values through the relevant {@link org.elasticsearch.index.mapper.ValueFetcher}
  * and returns them as document fields.
  */
 public final class FetchFieldsPhase implements FetchSubPhase {
+
+    private static final List<FieldAndFormat> METADATA_FIELDS = List.of(
+        new FieldAndFormat(IgnoredFieldMapper.NAME, null),
+        new FieldAndFormat(RoutingFieldMapper.NAME, null),
+        new FieldAndFormat(LegacyTypeFieldMapper.NAME, null)
+    );
+
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
-        FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
-        if (fetchFieldsContext == null) {
+        final FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
+        final StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+
+        // with _stored: _none_ we don't fetch metadata fields either (regardless of whether they are stored or not, for bwc reasons)
+        boolean fetchStoredFields = storedFieldsContext != null && storedFieldsContext.fetchFields();
+        if (fetchFieldsContext == null && fetchStoredFields == false) {
             return null;
         }
 
-        FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFieldsContext.fields());
+        final List<FieldAndFormat> fetchFields = fetchFieldsContext == null ? Collections.emptyList()
+            : fetchFieldsContext.fields() == null ? Collections.emptyList()
+            : fetchFieldsContext.fields();
+
+        final FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFields);
+        final FieldFetcher metadataFieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), METADATA_FIELDS);
```

Comment:
Sorry I read this multiple times but I am not sure I get what you means:
* which test is missing? one where both `fields` and `stored_fields` are none?
* I don't get your second comment can you clarify?

Reply:
> which test is missing? one where both fields and stored_fields are none? My understanding is that only stored_fields accepts "none" as an option.

No worries, let me clarify: one that requests some fields with `fields` and `stored_fields: _none_` . I believe that when fields are requested the fields context will be not null, hence the phase will execute, in which case you need to check again if fetching stored fields was disabled, or you end up returning the default metadata fields that should not be returned? We should test that and it surprises me that there are no failures. Does this make sense?

My second comment revolves around the fact that you use the stored fields context only to disable loading of default metadata fields, but other metadata fields requested via `stored_fields` are still retrieved via `StoredFieldsPhase`. Also, if you request `stored_fields: _ignored`, I believe that it's again retrieved from stored fields directly without going through value fetchers, which will break once you move the field to have doc_values. In a previous iteration you were adding metadata fields from stored fields context and retrieving them in FetchFieldsPhase. You also had a conditional in StoredFieldsPhase that excluded metadata fields from it. Without this latter conditional, you are now removing the isMetadataField flag from StoredField, but don't you still handle metadata fields in `StoredFieldsPhase` when they are explicitly requested? Those should go via value fetchers as well. I think that with the flag removed the may end up among document fields in the response, but I am not entirely sure, and weird that there are no test failures around it. Does this make sense?

Reply:
My understanding is that we can't fetch some metadata fields via FetchFieldsPhase because some metadata fields are missing a `ValueFetcher` implementation. Doing that would result in throwing an `UnsupportedOperationException`. Am I wrong?

`SeqNoFieldMapper` does this

```
public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
     throw new UnsupportedOperationException("Cannot fetch values for internal field [" + name() + "].");
}
```

(note also that it return a 5xx error...which is also an issue in serverless because that is mapped to a "Service Unavailble" error).